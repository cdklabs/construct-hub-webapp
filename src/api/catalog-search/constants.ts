// These are mapped to user-readable values to keep URL Params clean for users
export enum CatalogSearchSort {
  NameAsc = "asc",
  NameDesc = "desc",
  PublishDateAsc = "oldest",
  PublishDateDesc = "newest",
}
