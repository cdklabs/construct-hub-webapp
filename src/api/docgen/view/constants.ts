import * as reflect from "jsii-reflect";
import { Markdown } from "../render/markdown";
import { Transpile } from "../transpile/transpile";
import { Constant } from "./constant";

export class Constants {
  private readonly constants: Constant[];
  constructor(transpile: Transpile, properties: reflect.Property[]) {
    this.constants = properties
      .filter((p) => !p.protected && p.const)
      .map((p) => new Constant(transpile, p));
  }

  public render(): Markdown {
    if (this.constants.length === 0) {
      return Markdown.EMPTY;
    }

    const md = new Markdown({ header: { title: "Constants" } });
    for (const c of this.constants) {
      md.section(c.render());
    }
    return md;
  }
}
