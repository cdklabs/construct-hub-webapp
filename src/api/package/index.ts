export { createAssembly } from "./assemblies";
export { getAssetsPath, getFullPackageName } from "./util";
export { fetchMetadata } from "./metadata";
export type { Metadata } from "./metadata";
