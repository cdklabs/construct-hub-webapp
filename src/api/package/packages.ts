import { Language } from "../../constants/languages";
import { API_PATHS } from "../../constants/url";
import { Metadata } from "./metadata";

export interface Author {
  readonly name: string;
  readonly url: string;
}

export interface CatalogPackage {
  name: string;
  languages: Partial<Record<Language, Record<string, unknown>>>;
  version: string;
  description: string;
  author: Author;
  keywords: string[];
  metadata: Metadata;
}

export interface Packages {
  packages: CatalogPackage[];
}

/**
 * Fetch the catalog of all packages from the backend.
 */
export const fetchPackages = async (): Promise<Packages> => {
  const response = await fetch(API_PATHS.CATALOG_SUFFIX);

  if (!response.ok) {
    console.error(response.statusText);
    throw new Error(`Failed fetching packages index: ${response.statusText}`);
  }

  return response.json();
};
