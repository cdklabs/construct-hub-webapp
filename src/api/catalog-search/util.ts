import { CatalogSearchFilters, ExtendedCatalogPackage } from ".";
import { KEYWORD_IGNORE_LIST } from "../../constants/keywords";
import { Language } from "../../constants/languages";
import { CatalogPackage } from "../package/packages";
import { CatalogSearchSort } from "./constants";

type SortFunction = (
  p1: ExtendedCatalogPackage,
  p2: ExtendedCatalogPackage
) => number;

type FilterFunctionBuilder<T> = (
  filter: T
) => undefined | ((pkg: ExtendedCatalogPackage) => boolean);

const getStrSort = (isAscending: boolean): SortFunction => {
  return (p1, p2) => p1.name.localeCompare(p2.name) * (isAscending ? 1 : -1);
};

const getDateSort =
  (isAscending: boolean): SortFunction =>
  (p1, p2) => {
    const d1 = new Date(p1.metadata.date).getTime();
    const d2 = new Date(p2.metadata.date).getTime();

    if (d1 === d2) {
      return getStrSort(true)(p1, p2);
    }

    if (isAscending) {
      return d2 < d1 ? 1 : -1;
    }

    return d1 < d2 ? 1 : -1;
  };

const getDownloadsSort = (isAscending: boolean): SortFunction => {
  return (p1, p2) => {
    if (p1.downloads !== p2.downloads) {
      return (p1.downloads - p2.downloads) * (isAscending ? 1 : -1);
    } else {
      // break ties by alphabetical
      return getStrSort(!isAscending)(p1, p2);
    }
  };
};

const getLanguagesFilter: FilterFunctionBuilder<
  CatalogSearchFilters["languages"]
> = (languages) => {
  const languageSet =
    (languages?.length ?? 0) > 0 ? new Set(languages) : undefined;

  if (!languageSet || languageSet.has(Language.TypeScript)) {
    return undefined;
  }

  return (pkg) => {
    const isMatched = Object.keys(pkg.languages ?? {}).some((lang) =>
      languageSet.has(lang as Language)
    );

    return isMatched;
  };
};

const getCDKTypeFilter: FilterFunctionBuilder<CatalogSearchFilters["cdkType"]> =
  (cdkType) => {
    if (!cdkType) return undefined;
    return (pkg) => pkg.metadata?.constructFramework?.name === cdkType;
  };

const getCDKMajorFilter: FilterFunctionBuilder<
  CatalogSearchFilters["cdkMajor"]
> = (cdkMajor) => {
  if (typeof cdkMajor !== "number") return undefined;
  return (pkg) => pkg.metadata?.constructFramework?.majorVersion === cdkMajor;
};

const getKeywordsFilter: FilterFunctionBuilder<
  CatalogSearchFilters["keywords"]
> = (keywords) => {
  if (!keywords?.length) return undefined;

  return (pkg) => {
    const set = new Set<string>();

    for (const kw of pkg.keywords ?? []) {
      set.add(kw.toLocaleLowerCase());
    }

    for (const tag of pkg.metadata?.packageTags ?? []) {
      const label = tag.keyword?.label;
      if (label) {
        set.add(label.toLocaleLowerCase());
      }
    }

    for (const query of keywords) {
      if (set.has(query.toLocaleLowerCase())) {
        return true;
      }
    }

    return false;
  };
};

const getTagsFilter: FilterFunctionBuilder<CatalogSearchFilters["tags"]> = (
  tags
) => {
  if (!tags || !tags.length) {
    return undefined;
  }

  return (pkg) => {
    return (
      pkg.metadata?.packageTags?.some((tag) => {
        return tags.includes(tag.id);
      }) ?? false
    );
  };
};

export const SORT_FUNCTIONS: Record<CatalogSearchSort, SortFunction> = {
  [CatalogSearchSort.NameAsc]: getStrSort(true),
  [CatalogSearchSort.NameDesc]: getStrSort(false),
  [CatalogSearchSort.PublishDateAsc]: getDateSort(true),
  [CatalogSearchSort.PublishDateDesc]: getDateSort(false),
  [CatalogSearchSort.DownloadsAsc]: getDownloadsSort(true),
  [CatalogSearchSort.DownloadsDesc]: getDownloadsSort(false),
};

export const FILTER_FUNCTIONS: {
  [key in keyof Required<CatalogSearchFilters>]: FilterFunctionBuilder<
    CatalogSearchFilters[key]
  >;
} = {
  cdkType: getCDKTypeFilter,
  cdkMajor: getCDKMajorFilter,
  keywords: getKeywordsFilter,
  languages: getLanguagesFilter,
  tags: getTagsFilter,
};

/**
 * Returns a set of all the keywords associated with a package. This includes
 * publisher-based keywords and keywords from package tags.
 *
 * The set contains a single entry for each keyword, and all keywords are lowercased.
 *
 * Filters out all keywords that are in the ignore list.
 *
 * @param pkg The package
 * @returns The set of keywords
 */
export const renderAllKeywords = (pkg: CatalogPackage) => {
  const allKeywords = new Set<string>();
  const publisherKeywords = pkg.keywords ?? [];
  const tagKeywords = (pkg.metadata?.packageTags ?? [])
    .filter((t) => t.keyword)
    .map((t) => t.keyword!.label);

  for (const kw of [...publisherKeywords, ...tagKeywords]) {
    if (KEYWORD_IGNORE_LIST.has(kw)) {
      continue;
    }

    allKeywords.add(kw.toLocaleLowerCase());
  }

  return Array.from(allKeywords);
};
