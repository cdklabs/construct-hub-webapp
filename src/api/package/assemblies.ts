import * as spec from "@jsii/spec";
import * as reflect from "jsii-reflect";
import { getAssetsPath, getFullPackageName } from "./util";

type Assemblies = { [packageName: string]: spec.Assembly };

export async function createAssembly(
  name: string,
  version: string,
  scope?: string
): Promise<reflect.Assembly> {
  const assemblies = await fetchAssemblies(name, version, scope);
  const ts = new reflect.TypeSystem();
  Object.values(assemblies).forEach((a) => {
    ts.addAssembly(new reflect.Assembly(ts, a));
  });
  return ts.findAssembly(getFullPackageName(name, scope));
}

async function fetchAssemblies(
  name: string,
  version: string,
  scope?: string
): Promise<Assemblies> {
  const assemblies: Assemblies = {};
  const requested = new Set();

  async function recurse(_name: string, _version: string, _scope?: string) {
    const packageFqn = `${getFullPackageName(_name, _scope)}@${_version};`;

    if (requested.has(packageFqn)) {
      return;
    }

    requested.add(packageFqn);
    const assembly = await fetchAssembly(_name, _version, _scope);
    assemblies[packageFqn] = assembly;
    const promises: Promise<void>[] = [];

    for (const [d, v] of Object.entries(assembly.dependencies ?? {})) {
      const scopeAndName = d.split("/");
      if (scopeAndName.length > 1) {
        promises.push(recurse(scopeAndName[1], v, scopeAndName[0]));
      } else {
        promises.push(recurse(scopeAndName[0], v, ""));
      }
    }

    await Promise.all(promises);
  }

  await recurse(name, version, scope);

  return assemblies;
}

async function fetchAssembly(
  name: string,
  version: string,
  scope?: string
): Promise<spec.Assembly> {
  if (version.startsWith("^")) {
    version = version.substring(1, version.length);
  }

  const assemblyPath = `${getAssetsPath(name, version, scope)}/jsii.json`;
  const response = await fetch(assemblyPath);
  if (!response.ok) {
    throw new Error(
      `Failed fetching assembly for ${assemblyPath}: ${response.statusText}`
    );
  }
  return response.json();
}
