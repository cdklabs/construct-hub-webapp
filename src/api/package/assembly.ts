import * as spec from "@jsii/spec";
import { API_PATHS } from "../../constants/url";
import { getAssetsPath } from "./util";

/**
 * Fetch assembly of a specific package from the backend.
 */
export const fetchAssembly = async (
  name: string,
  version: string,
  scope?: string
): Promise<spec.Assembly> => {
  const assemblyPath = `${getAssetsPath(name, version, scope)}${
    API_PATHS.ASSEMBLY_SUFFIX
  }`;
  const response = await fetch(assemblyPath);
  if (!response.ok) {
    throw new Error(
      `Failed fetching assembly for ${assemblyPath}: ${response.statusText}`
    );
  }
  return response.json();
};
