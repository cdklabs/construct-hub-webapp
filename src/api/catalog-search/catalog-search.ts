import lunr from "lunr";
import { CDKType } from "../../constants/constructs";
import { Language } from "../../constants/languages";
import { CatalogPackage } from "../package/packages";
import { PackageStats } from "../stats";
import { CatalogSearchSort } from "./constants";
import { FILTER_FUNCTIONS, renderAllKeywords, SORT_FUNCTIONS } from "./util";

const INDEX_FIELDS = {
  AUTHOR_EMAIL: {
    name: "authorEmail",
    boost: 1,
  },
  AUTHOR_NAME: {
    name: "authorName",
    boost: 3,
  },
  DESCRIPTION: {
    name: "description",
    boost: 2,
  },
  KEYWORDS: {
    name: "keywords",
    boost: 2,
  },
  NAME: {
    name: "name",
    boost: 5,
  },
  PACKAGE_NAME: {
    name: "packageName",
    boost: 5,
  },
  SCOPE: {
    name: "scope",
    boost: 5,
  },
  TAG_NAMES: {
    name: "tagNames",
    boost: 2,
  },
} as const;

export interface ExtendedCatalogPackage extends CatalogPackage {
  authorEmail?: string;
  authorName?: string;
  downloads: number;
  id: string;
  packageName?: string;
  tagNames: string[];
  scope?: string;
}

export interface CatalogConstructFrameworks {
  [CDKType.awscdk]: CatalogConstructFrameworkMeta;
  [CDKType.cdktf]: CatalogConstructFrameworkMeta;
  [CDKType.cdk8s]: CatalogConstructFrameworkMeta;
}
export interface CatalogConstructFrameworkMeta {
  pkgCount: number;
  majorVersions: number[];
}

export interface CatalogSearchFilters {
  /**
   * The CDK Output Type to filter by. This functionality is not yet deployed on any BE so the implementation on the client is not final.
   */
  cdkType?: CDKType;
  /**
   * The CDK Type's major version to filter by. This param is ignored if no major is set
   */
  cdkMajor?: number;
  /**
   * A list of languages to filter by. Constructs that are not yet filtered out, will be
   * returned if they support any of the languages in this list.
   */
  languages?: Language[];
  /**
   * A list of keywords to filter by.
   */
  keywords?: string[];
  /**
   * A list of tags to filter by.
   */
  tags?: string[];
}

export type CatalogSearchResults = Map<string, ExtendedCatalogPackage>;

export interface CatalogSearchParams {
  query?: string;
  filters?: CatalogSearchFilters;
  sort?: CatalogSearchSort;
}

export class CatalogSearchAPI {
  private readonly map: CatalogSearchResults;
  private index: lunr.Index;
  /**
   * A map of detected keywords with a key representing the keyword, and a value representing the amount of occurences
   * the keyword has in the catalog.
   */
  public readonly keywords: Map<string, number>;
  /**
   * A map of detected Construct Frameworks which provides a count of libraries for that framework and a set of major versions detected
   */
  public readonly constructFrameworks: CatalogConstructFrameworks;

  constructor(catalogData: CatalogPackage[], stats: PackageStats) {
    const catalogMap = catalogData
      // Packages with the "construct-hub/hide-from-search" keyword are shadow-banned from search results
      .filter(
        (pkg) => !pkg.keywords?.includes("construct-hub/hide-from-search")
      )
      .reduce((map, pkg) => {
        const { author, name, version } = pkg;
        const id = [name, version].join("@");
        const downloads = stats.packages[name]?.downloads?.npm ?? 0;
        let [scope, packageName] = name.split("/");

        if (!packageName) {
          packageName = scope;
        }

        let authorName: string | undefined;
        let authorEmail: string | undefined;

        if (typeof author === "string") {
          authorName = author;
        } else {
          if (author?.name) {
            authorName = author.name;
          }

          if (author?.email) {
            authorEmail = author.email;
          }
        }
        map.set(id, {
          ...pkg,
          authorName,
          authorEmail,
          keywords: renderAllKeywords(pkg),
          downloads,
          id,
          packageName,
          scope,
          tagNames: (pkg.metadata.packageTags ?? []).map((t) => t.id),
        });

        return map;
      }, new Map<string, ExtendedCatalogPackage>());

    this.map = this.sort(catalogMap, CatalogSearchSort.PublishDateDesc);

    this.constructFrameworks = this.detectConstructFrameworks();
    this.keywords = this.detectKeywords();

    this.index = lunr(function () {
      this.tokenizer.separator = /[\s\-/@]+/;
      this.ref("id");

      for (const key in INDEX_FIELDS) {
        const field = INDEX_FIELDS[key as keyof typeof INDEX_FIELDS];
        this.field(field.name, { boost: field.boost });
      }

      [...catalogMap.values()].forEach((pkg) => {
        this.add(pkg);
      });
    });
  }

  /**
   * Performs a Search against the catalog and returns a Map with results ordered
   * by search score / relevance
   */
  public search(params?: CatalogSearchParams): CatalogSearchResults {
    const { query, filters, sort } = params ?? {};

    let results = query ? this.query(query) : new Map(this.map);

    // TODO: Investigate if we can leverage lunr for filtering
    if (filters) {
      results = this.filter(results, filters);
    }

    results = this.dedup(results);

    if (sort) {
      results = this.sort(results, sort);
    }

    return results;
  }

  /**
   * This calls the index search method and returns a map of results ordered by relevance.
   */
  private query(query: string): CatalogSearchResults {
    let refs: lunr.Index.Result[] = [];

    try {
      let tokenizedQuery = lunr.tokenizer(query);

      if (tokenizedQuery.length > 1) {
        // A large number of libraries include the term cdk within the title - which will lead to an
        // inflated result count. TODO: determine if there are other terms to filter out
        tokenizedQuery = tokenizedQuery.filter(
          (token) => token.toString() !== "cdk"
        );
      }

      refs = this.index.query((q) => {
        q.term(tokenizedQuery, {});
      });
    } catch (e) {
      console.error(e);
    }

    return refs.reduce((packages, { ref }) => {
      const pkg = this.map.get(ref);

      if (pkg) {
        packages.set(ref, pkg);
      }

      return packages;
    }, new Map() as CatalogSearchResults);
  }

  /**
   * Performs a Search against the catalog and returns an array of all packages matching the query.
   */
  public findByName(query: string): ExtendedCatalogPackage[] {
    const results = [...this.map.values()].filter((pkg) => pkg.name === query);
    const matches = new Array<ExtendedCatalogPackage>();
    for (const pkg of results.values()) {
      if (pkg.name === query) {
        matches.push(pkg);
      }
    }
    return matches;
  }

  /**
   * Filters query results. Mutates the passed-in map
   */
  private filter(
    results: CatalogSearchResults,
    filters: CatalogSearchFilters
  ): CatalogSearchResults {
    const { cdkType, cdkMajor, keywords, languages, tags } = filters;
    const copiedResults = new Map(results);

    const filterFunctions = [
      FILTER_FUNCTIONS.cdkType(cdkType),
      // Ignore major version filter if no CDK Type is defined
      FILTER_FUNCTIONS.cdkMajor(cdkType ? cdkMajor : undefined),
      FILTER_FUNCTIONS.keywords(keywords),
      FILTER_FUNCTIONS.languages(languages),
      FILTER_FUNCTIONS.tags(tags),
    ].filter(Boolean) as ((pkg: ExtendedCatalogPackage) => boolean)[];

    copiedResults.forEach((result) => {
      let isFiltered = false;

      filterFunctions.forEach((filterFn) => {
        if (!isFiltered && !filterFn(result)) {
          isFiltered = true;
        }
      });

      if (isFiltered) {
        copiedResults.delete(result.id);
      }
    });

    return copiedResults;
  }

  /**
   * Sort filtered results
   */
  private sort(
    results: CatalogSearchResults,
    strategy: CatalogSearchSort
  ): CatalogSearchResults {
    if (strategy) {
      return new Map(
        [...results.entries()].sort(([, p1], [, p2]) =>
          SORT_FUNCTIONS[strategy](p1, p2)
        )
      );
    } else {
      return results;
    }
  }

  /**
   * De-duplicates packages that appear multiple times, keeping the
   * most recently published one.
   */
  private dedup(results: CatalogSearchResults): CatalogSearchResults {
    const dedupedResults: Map<string, ExtendedCatalogPackage> = new Map();

    for (const [_key, pkg] of results) {
      const maybePkg = dedupedResults.get(pkg.name);

      if (
        !maybePkg ||
        new Date(maybePkg.metadata.date) < new Date(pkg.metadata.date)
      ) {
        dedupedResults.set(pkg.name, pkg);
      }
    }

    return dedupedResults;
  }

  /**
   * Creates a map of keywords with values representing the occurence of the keyword within the catalog.
   */
  private detectKeywords() {
    const results = [...this.map.values()].reduce(
      (keywords: Map<string, number>, pkg: ExtendedCatalogPackage) => {
        for (const keyword of renderAllKeywords(pkg)) {
          const entry = keywords.get(keyword);
          keywords.set(keyword, (entry ?? 0) + 1);
        }

        return keywords;
      },
      new Map<string, number>()
    );

    return results;
  }

  /**
   * Creates an object of found construct frameworks in the catalog map.
   * They are indexed by the name of the construct framework and record the count
   * of packages for that framework as well as a list of major versions.
   */
  private detectConstructFrameworks() {
    const results: CatalogConstructFrameworks = [...this.map.values()].reduce(
      (frameworks: CatalogConstructFrameworks, pkg: ExtendedCatalogPackage) => {
        const { metadata } = pkg;

        const frameworkName = metadata?.constructFramework?.name;
        const majorVersion = metadata?.constructFramework?.majorVersion;

        if (frameworkName) {
          const entry = frameworks[frameworkName];

          if (
            majorVersion !== undefined &&
            !entry.majorVersions.includes(majorVersion)
          ) {
            entry.majorVersions.push(majorVersion);
          }

          entry.pkgCount += 1;
        }

        return frameworks;
      },
      {
        [CDKType.awscdk]: {
          majorVersions: [],
          pkgCount: 0,
        },
        [CDKType.cdk8s]: {
          majorVersions: [],
          pkgCount: 0,
        },
        [CDKType.cdktf]: {
          majorVersions: [],
          pkgCount: 0,
        },
      }
    );

    return results;
  }
}
