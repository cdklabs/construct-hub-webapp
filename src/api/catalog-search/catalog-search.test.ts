import catalogFixture from "../../__fixtures__/catalog.json";
import { Language } from "../../constants/languages";
import { CatalogPackage } from "../package/packages";
import { CatalogSearchAPI } from "./catalog-search";
import { CatalogSearchSort } from "./constants";

describe("CatalogSearchAPI", () => {
  const instance = new CatalogSearchAPI(
    catalogFixture.packages as CatalogPackage[]
  );

  it("Returns all results for empty search query", () => {
    expect(instance.search().size).toEqual(catalogFixture.packages.length);
  });

  it("Returns results filtered by language", () => {
    const javaResults = instance.search({
      filters: { language: Language.Java },
    });

    expect(javaResults.size).toEqual(
      catalogFixture.packages.filter((p) => p.languages.java !== undefined)
        .length
    );

    const pythonResults = instance.search({
      filters: { language: Language.Python },
    });

    expect(pythonResults.size).toEqual(
      catalogFixture.packages.filter((p) => p.languages.python !== undefined)
        .length
    );
  });

  it("Returns results filtered by multiple languages", () => {
    const javaAndPythonResults = instance.search({
      filters: { languages: [Language.Java, Language.Python] },
    });

    expect(javaAndPythonResults.size).toEqual(
      catalogFixture.packages.filter(
        (p) =>
          p.languages.java !== undefined || p.languages.python !== undefined
      ).length
    );
  });

  it("Returns results ordered by Sort", () => {
    const publishDateAsc = [
      ...instance.search({ sort: CatalogSearchSort.PublishDateAsc }).values(),
    ];

    const publishDateDesc = [
      ...instance.search({ sort: CatalogSearchSort.PublishDateDesc }).values(),
    ];

    publishDateAsc.forEach(({ metadata: { date } }, index) => {
      expect(date).toEqual(
        publishDateDesc[publishDateDesc.length - 1 - index].metadata.date
      );
    });

    const nameAsc = [
      ...instance.search({ sort: CatalogSearchSort.NameAsc }).values(),
    ];
    const nameDesc = [
      ...instance.search({ sort: CatalogSearchSort.NameDesc }).values(),
    ];

    nameAsc.forEach(({ name }, index) => {
      expect(name).toEqual(nameDesc[nameDesc.length - 1 - index].name);
    });
  });

  describe("Snapshots", () => {
    it("Returns consistent query results", () => {
      const results = instance.search({
        query: "lambda libraries",
      });

      expect([...results].map(([id]) => id)).toMatchSnapshot();
    });

    it("Returns consistent filter results", () => {
      const results = instance.search({
        filters: {
          language: Language.Python,
        },
      });

      expect([...results].map(([id]) => id)).toMatchSnapshot();
    });

    it("Returns consistent sort results", () => {
      const results = instance.search({
        sort: CatalogSearchSort.NameAsc,
      });

      expect([...results].map(([id]) => id)).toMatchSnapshot();
    });
  });
});
