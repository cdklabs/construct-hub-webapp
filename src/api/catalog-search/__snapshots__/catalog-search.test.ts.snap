// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatalogSearchAPI Snapshots Returns consistent filter results 1`] = `
Array [
  "@aws-cdk/aws-backup",
  "@aws-cdk/aws-lambda-destinations",
  "@aws-cdk/aws-route53-patterns",
  "@aws-cdk/aws-ecs",
  "@aws-cdk/aws-apigatewayv2-authorizers",
  "@aws-cdk/aws-globalaccelerator-endpoints",
  "@aws-cdk/aws-dynamodb",
  "@aws-cdk/aws-ses",
  "@aws-cdk/aws-lambda-nodejs",
  "@aws-cdk/aws-kinesisfirehose",
  "@aws-cdk/aws-glue",
  "@aws-cdk/aws-servicecatalog",
  "@aws-cdk/aws-kinesis",
  "@aws-cdk/aws-synthetics",
  "@aws-cdk/aws-sqs",
  "@aws-cdk/aws-ecr-assets",
  "@aws-cdk/aws-ssm",
  "@aws-cdk/aws-ecr",
  "@aws-cdk/aws-workspaces",
  "@aws-cdk/aws-timestream",
  "@aws-cdk/aws-sso",
  "@aws-cdk/aws-securityhub",
  "@aws-cdk/aws-s3outposts",
  "@aws-cdk/aws-s3objectlambda",
  "@aws-cdk/aws-qldb",
  "@aws-cdk/aws-mediastore",
  "@aws-cdk/aws-mediapackage",
  "@aws-cdk/aws-medialive",
  "@aws-cdk/aws-mediaconnect",
  "@aws-cdk/aws-managedblockchain",
  "@aws-cdk/aws-location",
  "@aws-cdk/aws-iotwireless",
  "@aws-cdk/aws-iotcoredeviceadvisor",
  "@aws-cdk/aws-events",
  "@aws-cdk/aws-elasticbeanstalk",
  "@aws-cdk/aws-dms",
  "@aws-cdk/aws-dax",
  "@aws-cdk/aws-datasync",
  "@aws-cdk/aws-customerprofiles",
  "@aws-cdk/aws-codeguruprofiler",
  "@aws-cdk/aws-codeartifact",
  "@aws-cdk/aws-cloudwatch",
  "@aws-cdk/aws-ce",
  "@aws-cdk/aws-autoscalingplans",
  "@aws-cdk/aws-auditmanager",
  "@aws-cdk/aws-appstream",
  "@aws-cdk/aws-applicationinsights",
  "@aws-cdk/assets",
  "@aws-cdk/cx-api",
  "cdk-sns-notify",
  "cdk-tweet-queue",
  "cdk-alps-graph-ql",
  "gradle_s3_uploader",
  "cdk-mediapackage-custom",
  "cdk-noob",
  "cdk8s-aws-alb-ingress-controller",
  "cdk8s-plus",
  "cdk-iam-constants",
  "monocdk-experiment",
  "@rayou/cdk-url-shortener",
  "cdk-serverless-api",
  "cdk-stepfunctions-patterns",
  "aws-cdk-dynamodb-seeder",
  "@cdk-7layer-constructs/kinesis-firehose-transformer",
  "cdk-elasticache-monitor",
  "@taimos/cdk-construct-alexa-skill",
  "cdk-time-bomb",
  "cdk-athena-workgroup",
  "cdk-watchful-edge",
  "sqs-redrive",
  "cdk-github-import",
  "@philcali-cdk/static-website",
  "aws-lambda-golang",
  "cdk-sqs-monitored",
  "@moralesl/billing-alert",
  "sqs-redrive-construct",
  "cdk-secrets",
  "@terrastack/core",
  "cicd-spa-website",
  "aws-cdk-featureflags",
  "nodejsfunction-plus",
  "secure-bucket",
  "cdk-cicd",
  "@taimos/cdk-construct-gitlab-variable",
  "cdk-iam-generator",
  "@taimos/cdk-construct-hosting",
  "cdk-constants",
  "@aws-cdk/cdk",
]
`;

exports[`CatalogSearchAPI Snapshots Returns consistent query results 1`] = `
Array [
  "aws-lambda-golang",
  "@aws-cdk/aws-lambda-nodejs",
  "@aws-cdk/aws-lambda-destinations",
  "nodejsfunction-plus",
  "cdk-elasticache-monitor",
  "aws-cdk-featureflags",
  "@aws-cdk/aws-ssm",
  "@aws-cdk/aws-appstream",
  "@aws-cdk/aws-mediastore",
  "@aws-cdk/aws-ecs",
  "@aws-cdk/aws-ecr",
  "@aws-cdk/aws-backup",
  "@aws-cdk/aws-mediapackage",
  "@aws-cdk/aws-dms",
  "@aws-cdk/aws-ce",
  "@aws-cdk/aws-workspaces",
  "@aws-cdk/aws-synthetics",
  "@aws-cdk/aws-s3objectlambda",
  "@aws-cdk/aws-cloudwatch",
  "@aws-cdk/aws-location",
  "@aws-cdk/aws-securityhub",
  "@aws-cdk/aws-timestream",
  "@aws-cdk/aws-sqs",
  "@aws-cdk/aws-datasync",
  "@aws-cdk/aws-codeguruprofiler",
  "@aws-cdk/aws-sso",
  "@aws-cdk/aws-codeartifact",
  "@aws-cdk/aws-elasticbeanstalk",
  "@aws-cdk/aws-applicationinsights",
  "@aws-cdk/aws-servicecatalog",
  "@aws-cdk/aws-medialive",
  "@aws-cdk/aws-dynamodb",
  "@aws-cdk/aws-s3outposts",
  "@aws-cdk/aws-dax",
  "@aws-cdk/aws-kinesisfirehose",
  "@aws-cdk/aws-qldb",
  "@aws-cdk/aws-customerprofiles",
  "@aws-cdk/aws-mediaconnect",
  "@aws-cdk/aws-iotwireless",
  "@aws-cdk/aws-events",
  "@aws-cdk/aws-kinesis",
  "@aws-cdk/aws-iotcoredeviceadvisor",
  "@aws-cdk/aws-auditmanager",
  "@aws-cdk/aws-ses",
  "@aws-cdk/aws-autoscalingplans",
  "@aws-cdk/aws-managedblockchain",
  "@aws-cdk/aws-glue",
  "cdk-constants",
  "@aws-cdk/aws-route53-patterns",
  "aws-cdk-image-resize",
  "@pgarbe/cdk-bitbucket-server",
  "cdk8s-aws-alb-ingress-controller",
]
`;

exports[`CatalogSearchAPI Snapshots Returns consistent sort results 1`] = `
Array [
  "@aws-cdk/assets",
  "@aws-cdk/aws-apigatewayv2-authorizers",
  "@aws-cdk/aws-applicationinsights",
  "@aws-cdk/aws-appstream",
  "@aws-cdk/aws-auditmanager",
  "@aws-cdk/aws-autoscalingplans",
  "@aws-cdk/aws-backup",
  "@aws-cdk/aws-ce",
  "@aws-cdk/aws-cloudwatch",
  "@aws-cdk/aws-codeartifact",
  "@aws-cdk/aws-codeguruprofiler",
  "@aws-cdk/aws-customerprofiles",
  "@aws-cdk/aws-datasync",
  "@aws-cdk/aws-dax",
  "@aws-cdk/aws-dms",
  "@aws-cdk/aws-dynamodb",
  "@aws-cdk/aws-ecr",
  "@aws-cdk/aws-ecr-assets",
  "@aws-cdk/aws-ecs",
  "@aws-cdk/aws-elasticbeanstalk",
  "@aws-cdk/aws-events",
  "@aws-cdk/aws-globalaccelerator-endpoints",
  "@aws-cdk/aws-glue",
  "@aws-cdk/aws-iotcoredeviceadvisor",
  "@aws-cdk/aws-iotwireless",
  "@aws-cdk/aws-kinesis",
  "@aws-cdk/aws-kinesisfirehose",
  "@aws-cdk/aws-lambda-destinations",
  "@aws-cdk/aws-lambda-nodejs",
  "@aws-cdk/aws-location",
  "@aws-cdk/aws-managedblockchain",
  "@aws-cdk/aws-mediaconnect",
  "@aws-cdk/aws-medialive",
  "@aws-cdk/aws-mediapackage",
  "@aws-cdk/aws-mediastore",
  "@aws-cdk/aws-qldb",
  "@aws-cdk/aws-route53-patterns",
  "@aws-cdk/aws-s3objectlambda",
  "@aws-cdk/aws-s3outposts",
  "@aws-cdk/aws-securityhub",
  "@aws-cdk/aws-servicecatalog",
  "@aws-cdk/aws-ses",
  "@aws-cdk/aws-sqs",
  "@aws-cdk/aws-ssm",
  "@aws-cdk/aws-sso",
  "@aws-cdk/aws-synthetics",
  "@aws-cdk/aws-timestream",
  "@aws-cdk/aws-workspaces",
  "@aws-cdk/cdk",
  "@aws-cdk/cx-api",
  "@cdk-7layer-constructs/kinesis-firehose-transformer",
  "@moralesl/billing-alert",
  "@pgarbe/cdk-bitbucket-server",
  "@philcali-cdk/static-website",
  "@rayou/cdk-url-shortener",
  "@taimos/cdk-construct-alexa-skill",
  "@taimos/cdk-construct-gitlab-variable",
  "@taimos/cdk-construct-hosting",
  "@terrastack/core",
  "@umlss/aws-cdk-serverless-timer",
  "aws-cdk-dynamodb-seeder",
  "aws-cdk-featureflags",
  "aws-cdk-image-resize",
  "aws-lambda-golang",
  "cdk-alps-graph-ql",
  "cdk-athena-workgroup",
  "cdk-cicd",
  "cdk-cloudfront-deploy",
  "cdk-constants",
  "cdk-elasticache-monitor",
  "cdk-github-import",
  "cdk-iam-constants",
  "cdk-iam-generator",
  "cdk-image-builder-patterns",
  "cdk-libs",
  "cdk-mediapackage-custom",
  "cdk-noob",
  "cdk-secrets",
  "cdk-serverless-api",
  "cdk-sns-notify",
  "cdk-sqs-monitored",
  "cdk-stepfunctions-patterns",
  "cdk-time-bomb",
  "cdk-tweet-queue",
  "cdk-watchful-edge",
  "cdk-website",
  "cdk8s-aws-alb-ingress-controller",
  "cdk8s-op",
  "cdk8s-plus",
  "cicd-spa-website",
  "crossplane-cdk8s",
  "gradle_s3_uploader",
  "monocdk-experiment",
  "nodejsfunction-plus",
  "p82yf-libocr",
  "secure-bucket",
  "sqs-redrive",
  "sqs-redrive-construct",
]
`;
